src/js/components/AppListComponent.jsx

```diff
@@ -148,105 +148,135 @@ var AppListComponent = React.createClass({
     });
   },
 
-  getGroupedNodes: function (appsSequence) {
-    var currentGroup = this.props.currentGroup;
-
-    return appsSequence
-      .filter(app => app.id.startsWith(currentGroup))
-      .reduce((memo, app) => {
-        let relativePath = app.id.substring(currentGroup.length);
-        let pathParts = relativePath.split("/");
-        let isGroup = pathParts.length > 1;
-
-        if (!isGroup) {
-          memo.push(app);
-        } else {
-          let groupId = currentGroup + pathParts[0];
-          let group = memo.find(item => {
-            return item.id === groupId;
-          });
-
-          if (group == null) {
-            group = initGroupNode(groupId, app);
-            memo.push(group);
-          } else {
-            updateGroupNode(group, app);
-          }
-        }
-        return memo;
-      }, []);
+  hasFilters: function () {
+    return Object.values(this.props.filters).some((filter) => {
+      return (Util.isArray(filter) && filter.length > 0) ||
+        (Util.isString(filter) && filter != null && filter !== "");
+    });
   },
 
-  getAppNodes: function () {
-    var state = this.state;
-    var sortKey = state.sortKey;
+  filterNodes: function (nodesSequence) {
     var props = this.props;
+    var currentGroup = props.currentGroup;
+    var filters = props.filters;
 
-    var appsSequence = lazy(state.apps);
+    if (currentGroup !== "/") {
+      nodesSequence = nodesSequence
+        .filter(app => app.id.startsWith(currentGroup));
+    }
 
-    if (props.filterText != null && props.filterText !== "") {
-      appsSequence = appsSequence
+    if (filters.filterText != null && filters.filterText !== "") {
+      nodesSequence = nodesSequence
         .filter(function (app) {
-          return app.id.indexOf(props.filterText) !== -1;
+          return app.id.indexOf(filters.filterText) !== -1;
         });
     }
 
-    if (props.filterLabels != null && props.filterLabels.length > 0) {
-      appsSequence = appsSequence.filter(function (app) {
+    if (filters.filterLabels != null && filters.filterLabels.length > 0) {
+      nodesSequence = nodesSequence.filter(function (app) {
         let labels = app.labels;
         if (labels == null || Object.keys(labels).length === 0) {
           return false;
         }
 
-        return lazy(props.filterLabels).some(function (label) {
+        return lazy(filters.filterLabels).some(function (label) {
           let [key, value] = label;
           return labels[key] === value;
         });
       });
     }
 
-    if (props.filterStatus != null && props.filterStatus.length > 0) {
-      appsSequence = appsSequence.filter(function (app) {
+    if (filters.filterStatus != null && filters.filterStatus.length > 0) {
```

Comment:
We could, but that would be only a precaution. If filterStatus isn't an array, we made something wrong before. And the good thing, without an array check we will notice then. A js error will arise. With an array check we won't recognize our mistake directly.